### DISCUSSION TITLE
What's the the most Theian way to render 3D content in Mini-Browser/Webview Browser/Other?

### DISCUSSION THREAD

**Discussion by RodogInfinite** [eclipse-theia/theia#10271] [Q&A]
I've been digging for a while and found clues here and there, but I'm still very new to how Theia works under the hood and so I don't have a solid understanding of the path forward. 

# Goal
Ideally, I'd like to play around with WGPU to render content within a window displayed in my browser-based-Theia IDE. I'm not sure if it's as simple as having either Firefox Nightly/ Chrome Canary be the browser that has the IDE open and then having the content displayed in the mini-browser/webview using the [proper canvas rendering](https://www.w3.org/TR/webgpu/#canvas-rendering). That seems too easy to be true; however, if it's not, then trying to get [this example](https://gfx-rs.github.io/2020/04/21/wgpu-web.html) to run soon is my goal so I can move on to the 3D examples. My fallback for having a 3D example rendered is using OpenGL as a proof of concept in mini-browser/webviews.

# Current Progress
What I've been able to do so far is get the [WGPU example](https://gfx-rs.github.io/2020/04/21/wgpu-web.html) to run by starting a [simple-http-server](https://crates.io/crates/simple-http-server) in the `generated` directory created in the example instructions. Doing that will display content in Firefox Nightly/Chrome Canary (when all of the unstable parts decide to work together -- toss up each day), but this causes an issue because there's a server running within the server running the Theia browser and that causes unexpected behavior. For example, losing the triangle upon refresh and not getting it back until many refresh requests are sent. My thoughts for moving forward are to either somehow serve the content from the same server as Theia or somehow have the 3D content server running in parallel that would automatically start when Theia starts and then having mini-browser/webviews listen to that server. I'm not sure which is better to pursue and have little understanding of the consequences of each option.

# TL;DR: Questions
The current questions I have are:
1. What is the best way to serve that 3D content to either mini-browser/webviews from browser based Theia?
2. Which would be better for the task in the long run: mini-browser, webviews, other?

**Comment by paul-marechal** [comment]
I once made a [proof-of-concept Theia extension](https://github.com/paul-marechal/theia-3d-view) to visualize `.obj` files using [Three.js](https://threejs.org/). This work shall only be relied on as an example.

Ways I can think of to implement 3D rendering in a Theia application:

1. Make a VS Code extension that uses Webviews. Then if you get it to work in VS Code but encounter issues in Theia it most likely means we need to fix our implementation of their API. If it doesn't work in VS Code this way then I don't think we'll support it either.
2. Make a Theia extension that contributes a widget that directly integrates your renderer.

**Reply by RodogInfinite** [reply]
Thank you for your response. I will be digging into your example when I have time. I expect this project will take a while and I'd be happy to share any progress with the community as I go. 

I was considering both of those implementations you suggested, but I'm not exactly sure which would be the most time efficient, maintainable option since I have no experience with creating extensions on VS Code/Theia and I'm a novice, at best, with networking. 

- Since I have no experience with Typescript or extensions, would you mind informing me of the general flow of your extension? Does it use the same server as the browser instance to move the data from the .obj file to the window or are you running a secondary server to do that?
- One last thing for now, could you please add a license to your extension so that it's clear for if I use parts of it as a template for experimenting to eventually get to an end result?

**Reply by paul-marechal** [reply]
@RodogInfinite sorry for not responding earlier. I just did a big cleanup and licensed the repo under MIT :)
